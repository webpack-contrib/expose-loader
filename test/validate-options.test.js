import { getCompiler, compile } from './helpers';

describe('validate options', () => {
  const tests = {
    expose: {
      success: [
        'globalObject1',
        'globalObject1.foo',
        ['globalObject1'],
        ['globalObject1.foo'],
        ['globalObject1.foo', 'globalObject1.bar'],
      ],
      failure: [false, true, /test/, '', [], [''], {}],
    },
    unknown: {
      success: [],
      failure: [1, true, false, 'test', /test/, [], {}, { foo: 'bar' }],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === 'object' && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === 'success' ? 'successfully validate' : 'throw an error on'
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      const compiler = getCompiler('./simple-commonjs2-single-export.js', {
        [key]: value,
      });

      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === 'success') {
          expect(stats.hasErrors()).toBe(false);
        } else if (type === 'failure') {
          const {
            compilation: { errors },
          } = stats;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
