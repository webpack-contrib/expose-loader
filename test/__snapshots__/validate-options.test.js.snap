// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "expose" option with "" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options.expose should be an non-empty string."
`;

exports[`validate options should throw an error on the "expose" option with "/test/" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options.expose should be one of these:
   non-empty string | [non-empty string, ...] (should not have fewer than 1 item)
   Details:
    * options.expose should be a non-empty string.
    * options.expose should be an array:
      [non-empty string, ...] (should not have fewer than 1 item)"
`;

exports[`validate options should throw an error on the "expose" option with "[""]" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options.expose[0] should be an non-empty string."
`;

exports[`validate options should throw an error on the "expose" option with "[]" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options.expose should be an non-empty array."
`;

exports[`validate options should throw an error on the "expose" option with "{}" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options.expose should be one of these:
   non-empty string | [non-empty string, ...] (should not have fewer than 1 item)
   Details:
    * options.expose should be a non-empty string.
    * options.expose should be an array:
      [non-empty string, ...] (should not have fewer than 1 item)"
`;

exports[`validate options should throw an error on the "expose" option with "false" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options.expose should be one of these:
   non-empty string | [non-empty string, ...] (should not have fewer than 1 item)
   Details:
    * options.expose should be a non-empty string.
    * options.expose should be an array:
      [non-empty string, ...] (should not have fewer than 1 item)"
`;

exports[`validate options should throw an error on the "expose" option with "true" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options.expose should be one of these:
   non-empty string | [non-empty string, ...] (should not have fewer than 1 item)
   Details:
    * options.expose should be a non-empty string.
    * options.expose should be an array:
      [non-empty string, ...] (should not have fewer than 1 item)"
`;

exports[`validate options should throw an error on the "unknown" option with "/test/" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { expose? }"
`;

exports[`validate options should throw an error on the "unknown" option with "[]" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { expose? }"
`;

exports[`validate options should throw an error on the "unknown" option with "{"foo":"bar"}" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { expose? }"
`;

exports[`validate options should throw an error on the "unknown" option with "{}" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { expose? }"
`;

exports[`validate options should throw an error on the "unknown" option with "1" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { expose? }"
`;

exports[`validate options should throw an error on the "unknown" option with "false" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { expose? }"
`;

exports[`validate options should throw an error on the "unknown" option with "test" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { expose? }"
`;

exports[`validate options should throw an error on the "unknown" option with "true" value 1`] = `
"Invalid options object. Expose Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { expose? }"
`;
